<div style="max-width: 700px; margin: 20px auto 30px; padding: 20px; text-align: center; border-bottom: 2px solid #e5e7eb;">
  <a href="01b_arma_models.ipynb" style="display: inline-block; padding: 10px 24px; background: #f9fafb; color: #38549c; text-decoration: none; border-radius: 8px; font-weight: 600; font-size: 14px; border: 1px solid #e5e7eb;">
    ← Part II - ARMA Models
  </a>
</div>

<div style="max-width: 880px; margin: 20px auto 22px; padding: 0px; border-radius: 18px; border: 1px solid #e5e7eb; background: linear-gradient(180deg, #ffffff 0%, #f9fafb 100%); box-shadow: 0 8px 26px rgba(0,0,0,0.06); overflow: hidden;">

  <!-- Banner Header -->
  <div style="padding: 34px 32px 14px; text-align: center; line-height: 1.38;">
    <div style="font-size: 13px; letter-spacing: 0.14em; text-transform: uppercase; color: #6b7280; font-weight: bold; margin-bottom: 5px;">
      Session #1
    </div>
    <div style="font-size: 29px; font-weight: 800; color: #14276c; margin-bottom: 4px;">
      Univariate linear time series models
    </div>
    <div style="font-size: 29px; font-weight: 800; color: #14276c; margin-bottom: 4px;">
      Part III: Unit Roots
    </div>
    <div style="font-size: 16.5px; color: #374151; font-style: italic; margin-bottom: 0;">
      Advanced Training School: Methods for Time Series
    </div>
  </div>

  <!-- Logo Section -->
  <div style="background: none; text-align: center; margin: 30px 0 10px;">
    <img src="https://www.cemfi.es/images/Logo-Azul.png" alt="CEMFI Logo" style="width: 158px; filter: drop-shadow(0 2px 12px rgba(56,84,156,0.05)); margin-bottom: 0;">
  </div>

  <!-- Name -->
  <div style="font-family: 'Times New Roman', Times, serif; color: #38549c; text-align: center; font-size: 1.22em; font-weight: bold; margin-bottom: 0px;">
    Jesus Villota Miranda © 2025
  </div>

  <!-- Contact info -->
  <div style="font-family: 'Times New Roman', Times, serif; color: #38549c; text-align: center; font-size: 1em; margin-top: 7px; margin-bottom: 20px;">
    <a href="mailto:jesus.villota@cemfi.edu.es" style="color: #38549c; text-decoration: none; margin-right:8px;" title="Email">
      jesus.villota@cemfi.edu.es
    </a>
    <span style="color:#9fa7bd;">|</span>
    <a href="https://www.linkedin.com/in/jesusvillotamiranda/" target="_blank" style="color: #38549c; text-decoration: none; margin-left:7px;" title="LinkedIn">
      LinkedIn
    </a>
  </div>
</div>

<div style="max-width: 700px; margin: 20px auto 30px; padding: 24px 32px; border-radius: 18px; border: 1px solid #e5e7eb; background: linear-gradient(180deg, #ffffff 0%, #f9fafb 100%); box-shadow: 0 8px 26px rgba(0,0,0,0.06);">

<div style="font-size: 14px; letter-spacing: 0.1em; text-transform: uppercase; color: #6b7280; font-weight: bold; margin-bottom: 12px; text-align: center;">
  Session #1 • Part III
</div>

<div style="font-size: 22px; font-weight: 700; color: #14276c; margin-bottom: 16px; text-align: center;">
  Table of Contents
</div>

<div style="font-size: 15px; color: #374151; line-height: 1.8;">
  <div style="margin-bottom: 12px;"><strong>11.</strong> <a href="#section-11" style="color: #38549c; text-decoration: none;">Non-stationarity & Integration</a></div>
  <div style="margin-bottom: 12px;"><strong>12.</strong> <a href="#section-12" style="color: #38549c; text-decoration: none;">Dickey-Fuller Tests</a></div>
  <div style="margin-bottom: 12px;"><strong>13.</strong> <a href="#section-13" style="color: #38549c; text-decoration: none;">Phillips-Perron Test</a></div>
  <div style="margin-bottom: 12px;"><strong>14.</strong> <a href="#section-14" style="color: #38549c; text-decoration: none;">KPSS Test</a></div>
  <div style="margin-bottom: 12px;"><strong>15.</strong> <a href="#section-15" style="color: #38549c; text-decoration: none;">Putting It Together: Decision Logic</a></div>
  <div style="margin-bottom: 0px;"><strong>16.</strong> <a href="#section-16" style="color: #38549c; text-decoration: none;">Practical Checklist</a></div>
</div>

</div>

<style>
h2.styled-header {
    max-width: 600px;
    margin: 20px auto 22px !important;
    padding: 20px 32px !important;
    border-radius: 18px;
    border: 1px solid #e5e7eb;
    background: linear-gradient(180deg, #ffffff 0%, #f9fafb 100%);
    box-shadow: 0 8px 26px rgba(0,0,0,0.06);
    overflow: hidden;
    text-align: center;
    font-size: 20px !important;
    font-weight: 800 !important;
    color: #14276c !important;
    margin-bottom: 8px !important;
    margin-top: 0 !important;
}
</style>


In [25]:
* ===================================================
* Data Exploration: Summary Statistics and Checks
* ===================================================
* Before testing for unit roots, explore the data
* Summary statistics for level and first difference
summarize lneurusd d_lneurusd, detail

* Check for missing values
count if missing(lneurusd)
count if missing(d_lneurusd)

* Display time series structure
tsset date
tsset  // display time series information

* Note: We have 5,258 observations after accounting for first difference
* This is sufficient for reliable unit root testing (recommended: T > 100)




                  Log EUR/USD Exchange Rate
-------------------------------------------------------------
      Percentiles      Smallest
 1%    -.1510555      -.1899506
 5%    -.1012578      -.1898297
10%    -.0152659      -.1786483       Obs               6,730
25%     .0810269      -.1786483       Sum of wgt.       6,730

50%     .1571747                      Mean           .1592579
                        Largest       Std. dev.      .1318484
75%     .2619026       .4651796
90%     .3216849       .4652423       Variance        .017384
95%     .3645737       .4686277       Skewness      -.2322385
99%     .4426967       .4706284       Kurtosis       2.770172

               First Difference of Log EUR/USD
-------------------------------------------------------------
      Percentiles      Smallest
 1%     -.015197       -.030031
 5%    -.0095197      -.0267241
10%    -.0067176      -.0219574       Obs               5,258
25%    -.0031692      -.0219432       Sum of wgt.       5,258

In [26]:
* ===================================================
* Load EUR/USD Data and Set Up Time Series
* ===================================================
* Load processed EUR/USD exchange rate data
use "../../data/processed/eurusd_data.dta", clear

* Display basic information
describe
tsset date

* The dataset contains:
* - date: Daily date
* - eurusd: Exchange rate level (USD per EUR)
* - lneurusd: Log exchange rate (already computed)
* - d_lneurusd: First difference of log rate (returns, already computed)





Contains data from ../../data/processed/eurusd_data.dta
 Observations:         7,000                  
    Variables:             4                  5 Nov 2025 02:12
--------------------------------------------------------------------------------
Variable      Storage   Display    Value
    name         type    format    label      Variable label
--------------------------------------------------------------------------------
eurusd          float   %9.0g                 EUR/USD Exchange Rate
date            float   %td                   Date
lneurusd        float   %9.0g                 Log EUR/USD Exchange Rate
d_lneurusd      float   %9.0g                 First Difference of Log EUR/USD
--------------------------------------------------------------------------------
Sorted by: date


Time variable: date, 04jan1999 to 31oct2025, but with gaps
        Delta: 1 day


---

<a id="section-11"></a>

<h2 class="styled-header">11) Non-stationarity & Integration</h2>

### **11.1 Integration Order: I(0) vs I(1)**

A time series $Y_t$ is said to be **integrated of order $d$**, denoted $Y_t \sim I(d)$, if it requires $d$ differences to become stationary.

**Key Definitions:**

- **I(0)**: Stationary series (no differencing needed)
  - Mean and variance constant over time
  - Autocovariance depends only on lag, not time
  
- **I(1)**: Series that becomes stationary after **first differencing**
  - $Y_t \sim I(1)$ if $\Delta Y_t = Y_t - Y_{t-1} \sim I(0)$
  - Most common in economics: prices, exchange rates, GDP levels
  
- **I(d)**: General case requiring $d$ differences

**Economic Interpretation:**
- **I(0)**: Returns, growth rates, inflation (fluctuate around a constant mean)
- **I(1)**: Prices, exchange rates, GDP levels (wander without bound)

---

### **11.2 Random Walk Models**

The simplest **I(1)** process is the **random walk**:

**Random Walk (No Drift):**
$$Y_t = Y_{t-1} + \varepsilon_t, \quad \varepsilon_t \sim \text{i.i.d.}(0, \sigma^2)$$

**Properties:**
- $Y_t = Y_0 + \sum_{i=1}^{t} \varepsilon_i$ (cumulative sum of shocks)
- $\mathbb{E}[Y_t] = Y_0$ (constant mean)
- $\text{Var}(Y_t) = t\sigma^2$ ✗ **Non-stationary** (variance grows with time!)
- $\Delta Y_t = \varepsilon_t \sim I(0)$ ✓ **First difference is stationary**

**Random Walk with Drift:**
$$Y_t = \mu + Y_{t-1} + \varepsilon_t$$

**Properties:**
- $Y_t = Y_0 + \mu t + \sum_{i=1}^{t} \varepsilon_i$
- Mean: $\mathbb{E}[Y_t] = Y_0 + \mu t$ ✗ **Non-stationary** (trending mean)
- Variance: $\text{Var}(Y_t) = t\sigma^2$ ✗ **Non-stationary**
- $\Delta Y_t = \mu + \varepsilon_t \sim I(0)$ ✓ **First difference is stationary around drift**

---

### **11.3 Deterministic vs Stochastic Trends**

**Deterministic Trend:**
$$Y_t = \beta_0 + \beta_1 t + \varepsilon_t, \quad \varepsilon_t \sim I(0)$$

- Trend is **fixed** (deterministic function of time)
- Removing trend: $Y_t - \beta_1 t = \beta_0 + \varepsilon_t$ is stationary
- **Trend-stationary**: I(0) after detrending

**Stochastic Trend (Unit Root):**
$$Y_t = Y_{t-1} + \varepsilon_t = Y_0 + \sum_{i=1}^{t} \varepsilon_i$$

- Trend is **random** (cumulative sum of shocks)
- Removing deterministic trend **does not** make it stationary
- **Difference-stationary**: I(1), requires differencing

**Key Distinction:**
- **Deterministic trend**: Detrend → stationary
- **Stochastic trend**: Difference → stationary

**Why It Matters:**
- **Forecasting**: Deterministic trends predictable; stochastic trends not
- **Spurious regression**: Two I(1) series can appear correlated even if independent
- **Cointegration**: Two I(1) series may share a common stochastic trend

---

### **11.4 Visual Intuition: EUR/USD Exchange Rate**

The plots above show the EUR/USD exchange rate in levels and first differences. Notice how:
- The **level** (`lneurusd`) appears to wander without bound - a classic sign of non-stationarity
- The **first difference** (`d_lneurusd`) fluctuates around zero - suggesting stationarity

This visual inspection provides initial evidence that the exchange rate level may be I(1), while returns (first differences) are I(0). We will now test this formally using unit root tests.


In [27]:
* ===================================================
* Side-by-Side Comparison: Level vs First Difference
* ===================================================
* Create comparison plot to visualize non-stationarity vs stationarity
* This helps understand the difference between I(1) and I(0) series

* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Create side-by-side plots using graph combine
* First, save individual graphs
tsline lneurusd, ///
    title("Level: Log EUR/USD (I(1)?)") ///
    ytitle("Log Exchange Rate") ///
    name(level, replace) ///
    note("Non-stationary: wanders without bound")

tsline d_lneurusd, ///
    title("First Difference: Returns (I(0)?)") ///
    ytitle("First Difference") ///
    yline(0, lcolor(red) lpattern(dash)) ///
    name(diff, replace) ///
    note("Stationary: fluctuates around zero")

* Combine graphs side-by-side
graph combine level diff, ///
    title("Visual Comparison: Level vs First Difference", size(medium)) ///
    note("Left: Level appears non-stationary | Right: First difference appears stationary")

* This visual comparison provides initial evidence for unit root testing





Time variable: date, 04jan1999 to 31oct2025, but with gaps
        Delta: 1 day


file /Users/jesusvillotamiranda/.stata_kernel_cache/graph6.svg saved as SVG
    format
file /Users/jesusvillotamiranda/.stata_kernel_cache/graph6.pdf saved as PDF
    format


file /Users/jesusvillotamiranda/.stata_kernel_cache/graph7.svg saved as SVG
    format
file /Users/jesusvillotamiranda/.stata_kernel_cache/graph7.pdf saved as PDF
    format


file /Users/jesusvillotamiranda/.stata_kernel_cache/graph8.svg saved as SVG
    format
file /Users/jesusvillotamiranda/.stata_kernel_cache/graph8.pdf saved as PDF
    format


In [28]:
* ===================================================
* Visual Comparison: Level vs First Difference
* ===================================================
* Plot log level (likely non-stationary)
tsline lneurusd, ///
    title("EUR/USD: Log Exchange Rate (Level)") ///
    ytitle("Log Exchange Rate") ///
    note("Non-stationary: appears to wander without bound")

* Plot first difference (likely stationary)
tsline d_lneurusd, ///
    title("EUR/USD: First Difference (Returns)") ///
    ytitle("First Difference") ///
    yline(0, lcolor(red) lpattern(dash)) ///
    note("Stationary: fluctuates around zero")




file /Users/jesusvillotamiranda/.stata_kernel_cache/graph9.svg saved as SVG
    format
file /Users/jesusvillotamiranda/.stata_kernel_cache/graph9.pdf saved as PDF
    format


file /Users/jesusvillotamiranda/.stata_kernel_cache/graph10.svg saved as SVG
    format
file /Users/jesusvillotamiranda/.stata_kernel_cache/graph10.pdf saved as PDF
    format


---

<a id="section-12"></a>

<h2 class="styled-header">12) Dickey-Fuller Tests</h2>

### **12.1 The Dickey-Fuller (DF) Test**

The **Dickey-Fuller test** is the most common test for unit roots. It tests the null hypothesis that a series has a unit root (is I(1)) against the alternative that it is stationary (I(0)).

**Basic DF Regression:**

Consider an AR(1) process: $Y_t = \phi Y_{t-1} + \varepsilon_t$

**Null hypothesis** ($H_0$): $\phi = 1$ (unit root, non-stationary)

**Alternative hypothesis** ($H_1$): $|\phi| < 1$ (stationary)

**Test Regression:**
$$\Delta Y_t = (\phi - 1) Y_{t-1} + \varepsilon_t = \delta Y_{t-1} + \varepsilon_t$$

where $\delta = \phi - 1$.

**Test Statistic:**
$$t_{\delta} = \frac{\hat{\delta}}{se(\hat{\delta})}$$

**Critical Values:**
- ✗ **Non-standard distribution** (not Student's $t$!)
- ✗ Critical values are **more negative** than standard $t$-test
- ✗ Must use **Dickey-Fuller critical values** (or MacKinnon approximate $p$-values)

**Decision Rule:**
- If $t_{\delta} < \text{DF critical value}$: **Reject $H_0$** → Series is **stationary** (I(0))
- If $t_{\delta} > \text{DF critical value}$: **Fail to reject $H_0$** → Series has **unit root** (I(1))

---

### **12.2 Three Forms of the DF Test**

The DF test has three variants depending on deterministic components:

**1. No Constant, No Trend:**
$$\Delta Y_t = \delta Y_{t-1} + \varepsilon_t$$

- Use when: Series has zero mean and no trend
- Rare in practice

**2. With Constant (Drift):**
$$\Delta Y_t = \alpha + \delta Y_{t-1} + \varepsilon_t$$

- Use when: Series may have non-zero mean but no trend
- Most common for **levels** of economic variables

**3. With Constant and Trend:**
$$\Delta Y_t = \alpha + \beta t + \delta Y_{t-1} + \varepsilon_t$$

- Use when: Series may have deterministic trend
- Use when: Visual inspection suggests trending behavior

**Which Form to Use?**
- **Start with constant + trend** if series appears to trend
- **Use constant only** if no clear trend
- **Check sensitivity**: Results should be robust to specification

---

### **12.3 Augmented Dickey-Fuller (ADF) Test**

The **Augmented Dickey-Fuller (ADF)** test extends DF to handle **serial correlation** in $\varepsilon_t$.

**ADF Regression:**
$$\Delta Y_t = \alpha + \beta t + \delta Y_{t-1} + \sum_{i=1}^{p} \gamma_i \Delta Y_{t-i} + \varepsilon_t$$

**Key Addition:**
- **Lagged differences** $\Delta Y_{t-1}, \ldots, \Delta Y_{t-p}$ control for serial correlation
- Ensures $\varepsilon_t$ is white noise (required for valid inference)

**Lag Selection:**
- **Too few lags**: Residual autocorrelation → biased test
- **Too many lags**: Lower power (harder to reject $H_0$)

**Common Rules:**
1. **Schwert (1989)**: $p_{\max} = \lfloor 12(T/100)^{1/4} \rfloor$
2. **Information Criteria**: Choose $p$ that minimizes AIC or BIC
3. **Sequential Testing**: Start with $p_{\max}$, drop lags if insignificant

**Stata Default:**
- `dfuller` uses **Schwert rule** by default
- Can specify `lags(#)` manually

---

### **12.4 Interpreting ADF Test Results**

**Test Output Contains:**
- **Test statistic** ($t$-statistic on $\delta$)
- **Critical values** at 1%, 5%, 10% levels
- **MacKinnon approximate $p$-value** (preferred for inference)

**Decision:**
- **$p$-value < 0.05**: Reject $H_0$ → Series is **stationary** (I(0))
- **$p$-value ≥ 0.05**: Fail to reject $H_0$ → Series has **unit root** (I(1))

**Important Notes:**
- ADF has **low power** (hard to reject $H_0$ when true)
- **Failure to reject** does not prove unit root (may be low power)
- Always check **differenced series** to confirm I(1) classification


In [29]:
* ===================================================
* ADF Test: No Constant, No Trend
* ===================================================
* Test log EUR/USD exchange rate
* Form 1: No constant, no trend (rarely used in practice)
dfuller lneurusd, lags(0) noconstant

* Note: This form is rarely appropriate for economic data
* We include it for completeness, but typically use constant or constant+trend



Dickey–Fuller test for unit root        Number of obs  = 5,258
Variable: lneurusd                      Number of lags =     0

H0: Random walk without drift, a = 0, d = 0

                                       Dickey–Fuller
                   Test      -------- critical value ---------
              statistic           1%           5%          10%
--------------------------------------------------------------
 Z(t)            -1.225       -2.580       -1.950       -1.620
--------------------------------------------------------------


---

### **Practice Checkpoint: Understanding Unit Root Tests**

Before moving on, make sure you can answer:

1. **Why do we need unit root tests?** What happens if we use non-stationary data in regression models?

2. **What is the null hypothesis in ADF/PP tests?** What about KPSS tests? (Hint: They're opposite!)

3. **If ADF fails to reject H0 (p-value = 0.52), what does this mean?** 
   - Does it prove the series has a unit root?
   - What additional evidence should you gather?

4. **Why do we test the differenced series after testing the level?** What does this confirm?

**Answers:**
- Non-stationary data can lead to spurious regressions and invalid inference
- ADF/PP: H0 = unit root; KPSS: H0 = stationarity (opposite!)
- Failure to reject does NOT prove unit root (may be low power); always test differenced series
- Testing differenced series confirms I(1) classification: if level is I(1), difference should be I(0)

---


In [30]:
* ===================================================
* ADF Test: With Constant (Drift)
* ===================================================
* Form 2: With constant (most common for levels)
* Use default lag selection (Schwert rule)
dfuller lneurusd

* Interpretation:
* - H0: Unit root (non-stationary)
* - H1: Stationary
* - Check p-value: if < 0.05, reject H0 (stationary)
* - If p-value >= 0.05, fail to reject H0 (unit root likely)



Dickey–Fuller test for unit root        Number of obs  = 5,258
Variable: lneurusd                      Number of lags =     0

H0: Random walk without drift, d = 0

                                       Dickey–Fuller
                   Test      -------- critical value ---------
              statistic           1%           5%          10%
--------------------------------------------------------------
 Z(t)            -2.156       -3.430       -2.860       -2.570
--------------------------------------------------------------
MacKinnon approximate p-value for Z(t) = 0.2226.


In [ ]:
* ===================================================
* ADF Test: With Constant and Trend
* ===================================================
* Form 3: With constant and trend (use when series appears to trend)
dfuller lneurusd, trend

* This is often the most appropriate form for exchange rates
* which may exhibit trending behavior over long periods



Dickey–Fuller test for unit root        Number of obs  = 5,258
Variable: lneurusd                      Number of lags =     0

H0: Random walk with or without drift

                                       Dickey–Fuller
                   Test      -------- critical value ---------
              statistic           1%           5%          10%
--------------------------------------------------------------
 Z(t)            -2.137       -3.960       -3.410       -3.120
--------------------------------------------------------------
MacKinnon approximate p-value for Z(t) = 0.5250.


### **12.5 How to Read ADF Test Output**

When you run `dfuller`, Stata provides:

1. **Test Statistic (Z(t))**: The Dickey-Fuller t-statistic
   - Compare this to critical values
   - More negative values → stronger evidence against unit root

2. **Critical Values**: Provided at 1%, 5%, 10% significance levels
   - If test statistic < critical value → Reject H0 (stationary)
   - If test statistic > critical value → Fail to reject H0 (unit root)

3. **MacKinnon p-value**: Preferred for inference
   - p-value < 0.05 → Reject H0 → Stationary (I(0))
   - p-value ≥ 0.05 → Fail to reject H0 → Unit root (I(1))

**Example from output above:**
- Test statistic: -2.137
- 5% critical value: -3.410
- Since -2.137 > -3.410, we fail to reject H0
- p-value: 0.5250 > 0.05 → Unit root (I(1))

---


In [32]:
* ===================================================
* Lag Selection Using Information Criteria
* ===================================================
* Select optimal lag length for ADF test using AIC/BIC
* This is more formal than arbitrary lag choice

* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Method: Estimate ADF regression manually for different lag lengths
* and compare information criteria

* Calculate maximum lag using Schwert rule
quietly count
local T = r(N)
local pmax = floor(12*((`T'/100)^(1/4)))
display "Maximum lag (Schwert rule): " `pmax'

* Store AIC and BIC for different lag lengths
matrix ic_results = J(`pmax'+1, 3, .)
matrix colnames ic_results = Lags AIC BIC

* Test lag lengths from 0 to pmax
forvalues p = 0/`pmax' {
    quietly {
        * Estimate ADF regression manually
        gen L_lneurusd = L.lneurusd
        gen D_lneurusd = D.lneurusd
        
        * Create lagged differences
        forvalues i = 1/`p' {
            gen LD`i' = L`i'.D_lneurusd
        }
        
        * Estimate regression with trend
        if `p' > 0 {
            regress D_lneurusd L_lneurusd LD* _trend
        }
        else {
            regress D_lneurusd L_lneurusd _trend
        }
        
        * Extract AIC and BIC
        estat ic
        matrix ic = r(S)
        matrix ic_results[`p'+1, 1] = `p'
        matrix ic_results[`p'+1, 2] = ic[1,5]  // AIC
        matrix ic_results[`p'+1, 3] = ic[1,6]  // BIC
        
        * Clean up
        drop L_lneurusd D_lneurusd
        forvalues i = 1/`p' {
            capture drop LD`i'
        }
    }
}

* Display results
matrix list ic_results

* Find optimal lag (minimum AIC or BIC)
* Note: BIC is more conservative (prefers simpler models)
* For this demonstration, we'll use a simpler approach:

* Alternative simpler method: Use dfgls command if available
* or manually test a few lag lengths and compare

display "For practical purposes, test a few lag lengths:"
display "0, 4, 8, 12 lags and compare p-values"
display "Choose lag length where results are robust"
display "If p-values differ substantially, use information criteria above"





Time variable: date, 04jan1999 to 31oct2025, but with gaps
        Delta: 1 day




Maximum lag (Schwert rule): 34



variable _trend not found


r(111);
r(111);


In [ ]:
* ===================================================
* Residual Diagnostics After ADF Test
* ===================================================
* After running ADF test, check if residuals are white noise
* If residuals show autocorrelation, need more lags

* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Run ADF test and store residuals
* First, estimate ADF regression manually to get residuals
gen L_lneurusd = L.lneurusd
gen D_lneurusd = D.lneurusd
gen LD1 = L1.D_lneurusd
gen LD2 = L2.D_lneurusd
gen LD3 = L3.D_lneurusd
gen LD4 = L4.D_lneurusd

* Estimate ADF(4) regression with trend
regress D_lneurusd L_lneurusd LD* _trend

* Store residuals
predict residuals, residuals

* Test residuals for autocorrelation
* Ljung-Box Q test
wntestq residuals, lags(10)

* Interpretation:
* - H0: No autocorrelation (white noise)
* - H1: Autocorrelation present
* - If p-value < 0.05: Reject H0 → Need more lags
* - If p-value >= 0.05: Fail to reject H0 → Residuals are white noise ✓

* Plot ACF of residuals
ac residuals, lags(20) title("ACF of ADF Residuals") ///
    note("Should show no significant autocorrelation if lag length is adequate")

* Clean up
drop L_lneurusd D_lneurusd LD* residuals

* Note: If residuals show autocorrelation, increase lag length and re-test


---

### **Practice Checkpoint: Interpreting Test Results**

Test your understanding:

1. **ADF test gives p-value = 0.03. KPSS test statistic = 0.2 (critical value = 0.463).**
   - What is your conclusion? (I(0) or I(1)?)
   - Is this consistent evidence?

2. **ADF test gives p-value = 0.65. KPSS test statistic = 1.5 (critical value = 0.463).**
   - What is your conclusion?
   - Is this consistent evidence?

3. **What would you do if ADF and KPSS give conflicting results?**

**Answers:**
1. ADF rejects H0 (stationary), KPSS fails to reject H0 (stationary) → **Strong evidence of I(0)** ✓
2. ADF fails to reject H0 (unit root), KPSS rejects H0 (non-stationary) → **Strong evidence of I(1)** ✓
3. Investigate further: near-unit root, structural break, or low power. Test differenced series and check for breaks.

---


In [ ]:
* ===================================================
* ADF Test: Sensitivity to Lag Length
* ===================================================
* Test with different lag lengths to check robustness
* Too few lags: may have residual autocorrelation
* Too many lags: reduces power

* Ensure data is loaded and time series is set
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Test with 0 lags (simple DF)
display "ADF Test with 0 lags:"
dfuller lneurusd, lags(0) trend

* Test with 4 lags
display "ADF Test with 4 lags:"
dfuller lneurusd, lags(4) trend

* Test with 8 lags (more conservative, but reasonable for daily data)
display "ADF Test with 8 lags:"
dfuller lneurusd, lags(8) trend

* Compare results: Are conclusions robust to lag choice?
* If results differ, use information criteria to select optimal lag
* Note: For daily data, 8-12 lags is typically sufficient





Time variable: date, 04jan1999 to 31oct2025, but with gaps
        Delta: 1 day

ADF Test with 0 lags:


Dickey–Fuller test for unit root        Number of obs  = 5,258
Variable: lneurusd                      Number of lags =     0

H0: Random walk with or without drift

                                       Dickey–Fuller
                   Test      -------- critical value ---------
              statistic           1%           5%          10%
--------------------------------------------------------------
 Z(t)            -2.137       -3.960       -3.410       -3.120
--------------------------------------------------------------
MacKinnon approximate p-value for Z(t) = 0.5250.

ADF Test with 4 lags:

no observations


r(2000);
r(2000);


---

<a id="section-13"></a>

<h2 class="styled-header">13) Phillips-Perron Test</h2>

### **13.1 The Phillips-Perron (PP) Test**

The **Phillips-Perron test** is an alternative to ADF that uses **nonparametric corrections** for serial correlation and heteroskedasticity.

**Key Difference from ADF:**
- **ADF**: Adds **lagged differences** to control for serial correlation (parametric)
- **PP**: Uses **Newey-West correction** to adjust standard errors (nonparametric)

**PP Test Regression:**
$$\Delta Y_t = \alpha + \beta t + \delta Y_{t-1} + \varepsilon_t$$

**Same regression as DF**, but:
- **Standard errors** are adjusted using Newey-West HAC estimator
- **Test statistic** is modified to account for serial correlation
- **No lag selection needed** (nonparametric approach)

---

### **13.2 Advantages and Disadvantages**

**Advantages of PP:**
- ✓ **No lag selection** (avoids arbitrary choice of $p$)
- ✓ **Robust to heteroskedasticity** (uses HAC standard errors)
- ✓ **Same null/alternative** as ADF: $H_0$: unit root, $H_1$: stationary

**Disadvantages of PP:**
- ✗ **Bandwidth selection** needed for Newey-West (similar to lag selection)
- ✗ **Lower power** than ADF in some cases
- ✗ **Size distortions** in small samples

---

### **13.3 When to Use PP vs ADF?**

**Use ADF when:**
- Serial correlation structure is known/clear
- Want to model dynamics explicitly (lagged differences)
- Standard practice in most applications

**Use PP when:**
- Uncertain about lag structure
- Suspect heteroskedasticity
- Want robustness check on ADF results

**Best Practice:**
- **Report both tests** for robustness
- If **both reject** $H_0$: Strong evidence of stationarity
- If **both fail to reject**: Strong evidence of unit root
- If **conflicting results**: Investigate further (may indicate near-unit root or structural break)


In [ ]:
* ===================================================
* Phillips-Perron Test: With Constant
* ===================================================
* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Test log EUR/USD exchange rate using PP test
* PP uses nonparametric correction (no lag selection needed)
pperron lneurusd

* Interpretation:
* - Same null/alternative as ADF
* - H0: Unit root (non-stationary)
* - H1: Stationary
* - Compare p-value to ADF results for robustness
* - If p-value < 0.05: Reject H0 → Stationary (I(0))
* - If p-value >= 0.05: Fail to reject H0 → Unit root (I(1))


### **13.4 How to Read PP Test Output**

The Phillips-Perron test output is similar to ADF:

1. **Test Statistic (Z(t))**: Modified t-statistic with Newey-West correction
2. **Critical Values**: Same as ADF (non-standard distribution)
3. **MacKinnon p-value**: Use for inference

**Key Difference from ADF:**
- PP adjusts for serial correlation nonparametrically (no lag selection)
- Should give similar conclusions to ADF if both are correctly specified
- If results differ substantially, may indicate:
  - Heteroskedasticity (PP is more robust)
  - Near-unit root (different power properties)

**Interpretation:**
- p-value < 0.05 → Reject H0 → Stationary (I(0))
- p-value ≥ 0.05 → Fail to reject H0 → Unit root (I(1))

---


In [ ]:
* ===================================================
* Phillips-Perron Test: With Constant and Trend
* ===================================================
* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* PP test with trend (most appropriate for trending series)
pperron lneurusd, trend

* Compare with ADF results:
* - Do both tests give similar conclusions?
* - If conflicting, may indicate near-unit root or structural break
* - Expected: Both should fail to reject H0 (unit root) for exchange rate levels


---

<a id="section-14"></a>

<h2 class="styled-header">14) KPSS Test</h2>

### **14.1 The KPSS Test: A Stationarity Test**

The **Kwiatkowski-Phillips-Schmidt-Shin (KPSS)** test is **complementary** to ADF/PP:

**Key Difference:**
- **ADF/PP**: $H_0$: Unit root (I(1)) vs $H_1$: Stationary (I(0))
- **KPSS**: $H_0$: Stationary (I(0)) vs $H_1$: Unit root (I(1))

**KPSS tests the null of stationarity** (opposite of ADF/PP)!

**Why Use KPSS?**
- **Confirmation**: If ADF fails to reject $H_0$ (unit root), KPSS can confirm by rejecting its $H_0$ (non-stationary)
- **Power issues**: ADF has low power; KPSS may detect non-stationarity when ADF doesn't
- **Complementary evidence**: Both tests together provide stronger inference

---

### **14.2 KPSS Test Forms**

**KPSS Test Regression:**
$$Y_t = \alpha + \beta t + r_t + \varepsilon_t$$

where $r_t = r_{t-1} + u_t$ is a random walk component.

**Null hypothesis** ($H_0$): $\text{Var}(u_t) = 0$ → $r_t$ is constant → $Y_t$ is stationary

**Alternative hypothesis** ($H_1$): $\text{Var}(u_t) > 0$ → $r_t$ is random walk → $Y_t$ has unit root

**Two Forms:**

1. **Level Stationary** (`notrend`):
   - $Y_t = \alpha + r_t + \varepsilon_t$ (no trend)
   - Tests stationarity around a constant mean

2. **Trend Stationary** (`trend`):
   - $Y_t = \alpha + \beta t + r_t + \varepsilon_t$ (with trend)
   - Tests stationarity around a deterministic trend

---

### **14.3 Interpreting KPSS Results**

**Test Statistic:**
- KPSS statistic (LM-type test)
- **Large values** → reject $H_0$ (non-stationary)
- **Small values** → fail to reject $H_0$ (stationary)

**Critical Values:**
- Provided at 1%, 5%, 10% levels
- **If test statistic > critical value**: Reject $H_0$ → Series is **non-stationary** (I(1))
- **If test statistic < critical value**: Fail to reject $H_0$ → Series is **stationary** (I(0))

**Decision Logic:**
- **ADF fails to reject** (unit root) **AND KPSS rejects** (non-stationary) → **Strong evidence of I(1)**
- **ADF rejects** (stationary) **AND KPSS fails to reject** (stationary) → **Strong evidence of I(0)**
- **Conflicting results** → Investigate further (near-unit root, structural break, or low power)

---

### **14.4 Caveats and Limitations**

**Bandwidth Selection:**
- KPSS requires **bandwidth parameter** for Newey-West correction
- Default in Stata: Automatic selection
- Can be sensitive to bandwidth choice

**Size Distortions:**
- KPSS can have **size problems** (reject $H_0$ too often) in small samples
- Use with caution for $T < 100$

**Complementary Use:**
- **Never use KPSS alone**; always combine with ADF/PP
- Use as **confirmation** of ADF/PP results


---

### **Decision Flowchart: Unit Root Testing Workflow**

```
START: Load data and plot series
  │
  ├─→ Visual inspection suggests trend?
  │     │
  │     YES → Use ADF/PP with constant + trend
  │     NO  → Use ADF/PP with constant only
  │
  ├─→ Run ADF test (with appropriate specification)
  ├─→ Run PP test (robustness check)
  ├─→ Run KPSS test (complementary evidence)
  │
  ├─→ Compare results using decision matrix
  │
  ├─→ Are results consistent?
  │     │
  │     YES → Test differenced series to confirm
  │     NO  → Investigate: structural breaks? near-unit root?
  │
  ├─→ Test differenced series
  │     │
  │     If level is I(1) → Difference should be I(0) ✓
  │
  └─→ Final decision: Use levels or differences for modeling
```

**Key Principle:** Always combine multiple tests and verify by testing differenced series!

---


### **14.5 How to Read KPSS Test Output**

**Important:** KPSS has the OPPOSITE null hypothesis from ADF/PP!

**KPSS Output Contains:**
1. **Test Statistic**: LM-type statistic (larger = more evidence of non-stationarity)
2. **Critical Values**: At 1%, 5%, 10% levels
3. **Decision Rule**: 
   - If test statistic > critical value → Reject H0 → Non-stationary (I(1))
   - If test statistic < critical value → Fail to reject H0 → Stationary (I(0))

**Example Interpretation:**
- Test statistic: 2.5
- 5% critical value: 0.463
- Since 2.5 > 0.463, reject H0 → Non-stationary (I(1))

**Combining with ADF/PP:**
- ADF fails to reject (unit root) AND KPSS rejects (non-stationary) → Strong evidence of I(1) ✓
- ADF rejects (stationary) AND KPSS fails to reject (stationary) → Strong evidence of I(0) ✓

---


In [ ]:
* ===================================================
* KPSS Test: Level Stationary
* ===================================================
* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Note: KPSS test may require installation from SSC
* If command not found, install with: ssc install kpss, replace

* Test for level stationarity (no trend)
* H0: Stationary (I(0))
* H1: Non-stationary (I(1))
capture kpss lneurusd, notrend

* If kpss command not available, uncomment the following:
* ssc install kpss, replace
* kpss lneurusd, notrend

* Interpretation:
* - If test statistic > critical value: Reject H0 → Non-stationary (I(1))
* - If test statistic < critical value: Fail to reject H0 → Stationary (I(0))
* - For exchange rate levels, expect to reject H0 (non-stationary)


In [ ]:
* ===================================================
* KPSS Test: Trend Stationary
* ===================================================
* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Test for trend stationarity (with deterministic trend)
* H0: Trend stationary (I(0) after detrending)
* H1: Non-stationary (I(1))
capture kpss lneurusd, trend

* Compare with level stationary test
* For trending series, trend-stationary form is more appropriate
* Expected: Reject H0 (non-stationary) for exchange rate levels


---

<a id="section-15"></a>

<h2 class="styled-header">15) Putting It Together: Decision Logic</h2>

### **15.1 Comparing Test Results**

To determine the integration order of a series, we combine evidence from multiple tests:

**Decision Matrix:**

| ADF/PP Result | KPSS Result | Conclusion |
|---------------|-------------|------------|
| **Fail to reject** $H_0$ (unit root) | **Reject** $H_0$ (non-stationary) | ✓ **I(1)** - Strong evidence |
| **Reject** $H_0$ (stationary) | **Fail to reject** $H_0$ (stationary) | ✓ **I(0)** - Strong evidence |
| **Fail to reject** $H_0$ (unit root) | **Fail to reject** $H_0$ (stationary) | ⚠ **Uncertain** - Low power? |
| **Reject** $H_0$ (stationary) | **Reject** $H_0$ (non-stationary) | ⚠ **Conflicting** - Investigate! |

**Best Practice:**
1. Run **ADF with constant + trend**
2. Run **PP with constant + trend**
3. Run **KPSS with trend**
4. **Reconcile results** using decision matrix
5. **Confirm** by testing differenced series

---

### **15.2 Verifying I(1) Classification**

If level series is **I(1)**, then **first difference should be I(0)** (stationary).

**Verification Steps:**
1. Test **differenced series** with ADF/PP
   - Should **reject** $H_0$ (stationary)
2. Check **ACF/PACF** of differenced series
   - Should show **no persistent autocorrelation**
3. Test **residuals** for white noise
   - Ljung-Box Q test should **fail to reject** $H_0$

If differenced series is **stationary**, this confirms I(1) classification of level.


In [ ]:
* ===================================================
* Verify I(1) Classification: Test Differenced Series
* ===================================================
* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* If lneurusd is I(1), then d_lneurusd should be I(0) (stationary)
* Test first difference with ADF
dfuller d_lneurusd, lags(0) trend

* Expected result: Should REJECT H0 (stationary)
* This confirms that level series is I(1)
* If p-value < 0.05: Reject H0 → Differenced series is stationary → Level is I(1) ✓


In [ ]:
* ===================================================
* ACF and PACF of Differenced Series
* ===================================================
* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Check autocorrelation structure of differenced series
* Should show no persistent autocorrelation (stationary behavior)
ac d_lneurusd, lags(20) title("ACF: First Difference of Log EUR/USD") ///
    note("Should show no significant autocorrelation if stationary")

pac d_lneurusd, lags(20) title("PACF: First Difference of Log EUR/USD") ///
    note("Should show no significant partial autocorrelation if stationary")

* Interpretation:
* - If ACF/PACF show no significant spikes: Confirms stationarity
* - If significant autocorrelation: May need ARMA model for differenced series


In [ ]:
* ===================================================
* White Noise Test on Differenced Series
* ===================================================
* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Test if differenced series is white noise (no autocorrelation)
* Ljung-Box Q test
wntestq d_lneurusd, lags(10)

* Interpretation:
* - H0: White noise (no autocorrelation)
* - H1: Autocorrelation present
* - If p-value < 0.05: Reject H0 (autocorrelation present - may need ARMA model)
* - If p-value >= 0.05: Fail to reject H0 (white noise - fully stationary)
* - For exchange rate returns, may find some autocorrelation (common in financial data)


### **15.3 Summary Table: EUR/USD Exchange Rate**

Based on the tests above, we can summarize the results:

| Test | Specification | Test Statistic | p-value | Conclusion |
|------|---------------|----------------|---------|------------|
| **ADF** | Constant + Trend | Z(t) = -2.137 | 0.5250 | **I(1)** - Fail to reject H0 (unit root) |
| **PP** | Constant + Trend | [Run test] | [Check output] | [Compare with ADF] |
| **KPSS** | Trend | [Run test] | [Check output] | [Should reject H0 → I(1)] |
| **ADF (diff)** | Constant + Trend | [Run test] | [Check output] | **I(0)** - Should reject H0 (stationary) |

**Note:** Run all tests in the cells above to populate this table with your actual results.

**Typical Finding for Exchange Rates:**
- **Level** (`lneurusd`): I(1) - Unit root (non-stationary)
- **First Difference** (`d_lneurusd`): I(0) - Stationary (returns)

**Economic Interpretation:**
- Exchange rate levels follow a **random walk** (cannot predict future level)
- Exchange rate **returns** are stationary (can model with ARMA)
- Consistent with **efficient market hypothesis** (unpredictable levels, but returns may have short-term dependence)

**Decision:** Based on ADF test, EUR/USD exchange rate level is **I(1)**. Use first differences (returns) for modeling.


### **15.4 Structural Break Testing**

Unit root tests assume constant parameters over time. **Structural breaks** (regime changes) can cause:
- False rejection of H0 (spurious stationarity)
- False failure to reject H0 (spurious unit root)

**Common Structural Breaks:**
- Financial crises (2008, 2020)
- Policy regime changes
- Currency devaluations

**Testing for Structural Breaks:**

1. **Visual Inspection**: Plot series with vertical lines at potential break dates
2. **Split-Sample Tests**: Test pre- and post-break periods separately
3. **Zivot-Andrews Test**: Allows for break under H0 (more advanced)

---


In [ ]:
* ===================================================
* Structural Break Testing: Visual Inspection
* ===================================================
* Plot series with potential break dates marked
* Common breaks: 2008 financial crisis, 2020 COVID-19

* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Create break date markers
gen break2008 = td(15sep2008)  // Lehman Brothers collapse
gen break2020 = td(11mar2020)  // COVID-19 declared pandemic

* Plot with break dates
tsline lneurusd, ///
    title("EUR/USD: Log Exchange Rate with Structural Breaks") ///
    ytitle("Log Exchange Rate") ///
    xline(`=break2008', lcolor(red) lpattern(dash) lwidth(thin)) ///
    xline(`=break2020', lcolor(blue) lpattern(dash) lwidth(thin)) ///
    text(0.3 `=break2008' "2008 Crisis", size(small)) ///
    text(0.3 `=break2020' "2020 COVID", size(small)) ///
    note("Vertical lines indicate potential structural breaks")

* Visual inspection: Do you see obvious changes in level or trend?
* If yes, consider split-sample tests


In [ ]:
* ===================================================
* Structural Break Testing: Split-Sample ADF Tests
* ===================================================
* Test for unit root in pre- and post-break periods separately
* This checks if conclusions are robust to structural breaks

* Ensure data is loaded
use "../../data/processed/eurusd_data.dta", clear
tsset date

* Define break date (e.g., 2008 financial crisis)
gen break_date = td(15sep2008)

* Pre-break period (before 2008)
preserve
keep if date < break_date
tsset date
display "ADF Test: Pre-2008 Period"
dfuller lneurusd, lags(0) trend
restore

* Post-break period (after 2008)
preserve
keep if date >= break_date
tsset date
display "ADF Test: Post-2008 Period"
dfuller lneurusd, lags(0) trend
restore

* Interpretation:
* - If both periods show same conclusion (both I(1)): Robust to break
* - If conclusions differ: Structural break may affect inference
* - Consider using Zivot-Andrews test for formal break testing

* Note: Zivot-Andrews test requires installation:
* ssc install zandrews, replace
* za lneurusd, trend
* This test allows for break under H0 and is more appropriate for structural breaks


### **15.4 Optional: Replication on S&P 500 Prices**

As a reinforcement exercise, we can test S&P 500 prices (another classic I(1) series):


In [ ]:
* ===================================================
* Optional: Test S&P 500 Prices for Unit Root
* ===================================================
* Load S&P 500 data
use "../../data/processed/sp500_data.dta", clear
tsset date

* Display data structure
describe
summarize sp500_close sp500_return, detail

* Create log price if not already in dataset
capture confirm variable ln_sp500
if _rc {
    gen ln_sp500 = ln(sp500_close)
}

* Visual inspection: Plot level and returns
tsline ln_sp500, title("S&P 500: Log Price Level") ///
    ytitle("Log Price") note("Non-stationary: appears to wander")
    
tsline sp500_return, title("S&P 500: Returns") ///
    ytitle("Returns") yline(0, lcolor(red) lpattern(dash)) ///
    note("Stationary: fluctuates around zero")

* ADF test on log price level (with trend)
display "ADF Test on S&P 500 Log Price Level:"
dfuller ln_sp500, lags(0) trend

* Interpretation:
* - Expected: Fail to reject H0 (unit root) → I(1)
* - p-value >= 0.05 indicates unit root

* Test returns (already computed as sp500_return)
* Returns should be I(0) (stationary)
display "ADF Test on S&P 500 Returns:"
dfuller sp500_return, lags(0)

* Interpretation:
* - Expected: Reject H0 (stationary) → I(0)
* - p-value < 0.05 indicates stationarity
* - This confirms that prices are I(1) and returns are I(0)

* Summary:
* - Stock prices: I(1) - Unit root (non-stationary)
* - Stock returns: I(0) - Stationary
* - Consistent with efficient market hypothesis


---

<a id="section-16"></a>

<h2 class="styled-header">16) Practical Checklist</h2>

### **16.1 Deterministic Component Selection**

**Guidelines:**
1. **Visual inspection**: Plot the series first
   - Clear upward/downward trend → Use **constant + trend**
   - Fluctuates around constant level → Use **constant only**
   - Appears to wander → Use **constant + trend** (most conservative)

2. **Economic theory**: 
   - Prices, exchange rates → Usually **constant + trend**
   - Returns, growth rates → Usually **constant only**

3. **Sensitivity check**: 
   - Test with both specifications
   - If results differ, report both and discuss

---

### **16.2 Lag Selection for ADF**

**Methods:**
1. **Schwert Rule** (default in Stata):
   $$p_{\max} = \lfloor 12(T/100)^{1/4} \rfloor$$
   - Automatic, data-driven
   - Good starting point

2. **Information Criteria**:
   - Estimate ADF with different $p$
   - Choose $p$ that minimizes **AIC** or **BIC**
   - More formal, but computationally intensive

3. **Sequential Testing**:
   - Start with $p_{\max}$
   - Drop lags if $t$-statistic on last lag < 1.96
   - Stop when all lags significant

4. **Residual Diagnostics**:
   - After ADF, test residuals for autocorrelation
   - If significant autocorrelation, increase $p$

**Best Practice:**
- Use **Schwert rule** as default
- Check **sensitivity** to lag choice
- Report results for multiple lag lengths if conclusions differ

---

### **16.3 Structural Breaks**

**Problem:**
- Unit root tests assume **constant parameters** over time
- **Structural breaks** (regime changes) can cause:
  - False rejection of $H_0$ (spurious stationarity)
  - False failure to reject $H_0$ (spurious unit root)

**Solutions:**
1. **Visual inspection**: Look for obvious breaks in plots
2. **Chow test**: Test for parameter stability
3. **Split sample**: Test pre- and post-break separately
4. **Advanced tests**: Zivot-Andrews test (allows for break under $H_0$)

**Common Breaks:**
- Financial crises (2008, 2020)
- Policy regime changes
- Currency devaluations

---

### **16.4 Finite-Sample Power**

**Power Problem:**
- Unit root tests have **low power** (hard to reject $H_0$ when false)
- Especially problematic for:
  - **Near-unit roots** ($\phi \approx 0.95$)
  - **Small samples** ($T < 100$)
  - **High persistence** (slow mean reversion)

**Implications:**
- **Failure to reject** $H_0$ does **not prove** unit root
- May be **low power** (true $\phi = 0.95$, but test can't detect)
- Always check **differenced series** to confirm I(1)

**Improving Power:**
- Use **longer samples** when possible
- Combine **multiple tests** (ADF + PP + KPSS)
- Use **panel unit root tests** if multiple series available

---

### **16.5 Outliers and Data Quality**

**Issumes:**
- **Outliers** can affect test statistics
- **Missing values** can reduce effective sample size
- **Data errors** can create spurious patterns

**Pre-Testing Checks:**
1. **Summary statistics**: Check for extreme values
2. **Plots**: Visual inspection for outliers
3. **Missing data**: Handle appropriately (interpolation vs. deletion)
4. **Data quality**: Verify source and transformations

**Robustness:**
- Test with and without outliers
- Compare results across different sample periods
- Use **robust standard errors** when appropriate

---

### **16.6 Recommended Workflow**

**Step-by-Step Checklist:**

1. ✓ **Load data** and set time series structure (`tsset`)
2. ✓ **Plot series** (level and first difference)
3. ✓ **Visual inspection**: Identify trends, breaks, outliers
4. ✓ **ADF test**: Constant + trend, multiple lag lengths
5. ✓ **PP test**: Constant + trend (robustness check)
6. ✓ **KPSS test**: Trend (complementary evidence)
7. ✓ **Reconcile results**: Use decision matrix
8. ✓ **Test differenced series**: Confirm I(1) classification
9. ✓ **Residual diagnostics**: ACF, PACF, white noise tests
10. ✓ **Report results**: Table with all test statistics and $p$-values

**Final Decision:**
- If **consistent evidence** of I(1): Use **first differences** for modeling
- If **consistent evidence** of I(0): Use **levels** for modeling
- If **conflicting evidence**: Investigate further (breaks, near-unit root, low power)


---

## Summary of Part III

In this notebook, we have covered:

1. **Non-stationarity & Integration**: I(0) vs I(1) processes, random walks, deterministic vs stochastic trends
2. **Dickey-Fuller Tests**: DF and ADF tests with different deterministic specifications
3. **Phillips-Perron Test**: Nonparametric alternative to ADF
4. **KPSS Test**: Complementary stationarity test
5. **Decision Logic**: How to combine multiple tests for robust inference
6. **Practical Checklist**: Guidelines for proper implementation

**Key Takeaways:**
- **Unit root tests** are essential for determining integration order
- **Multiple tests** should be used together for robust inference
- **ADF/PP** test $H_0$: unit root; **KPSS** tests $H_0$: stationarity
- **Low power** is a common problem; failure to reject does not prove unit root
- **Always verify** I(1) classification by testing differenced series
- **Practical considerations** (lags, trends, breaks) are crucial for valid inference

**Next Steps:**
- In Part IV (if applicable), we will study **multivariate models** (VAR, cointegration)
- For now, we have the tools to properly test for unit roots and determine appropriate transformations

---

<div style="max-width: 700px; margin: 40px auto 20px; padding: 20px; text-align: center; border-top: 2px solid #e5e7eb;">
  <a href="01b_arma_models.ipynb" style="display: inline-block; padding: 10px 24px; background: #f9fafb; color: #38549c; text-decoration: none; border-radius: 8px; font-weight: 600; font-size: 14px; border: 1px solid #e5e7eb; margin-right: 12px;">
    ← Part II - ARMA Models
  </a>
</div>


<a id="section-11"></a>

<h2 class="styled-header">11) Non-stationarity & Integration</h2>

### **11.1 Integration Order: I(0) vs I(1)**

A time series $Y_t$ is said to be **integrated of order $d$**, denoted $Y_t \sim I(d)$, if it requires $d$ differences to become stationary.

**Key Definitions:**

- **I(0)**: Stationary series (no differencing needed)
  - Mean and variance constant over time
  - Autocovariance depends only on lag, not time
  
- **I(1)**: Series that becomes stationary after **first differencing**
  - $Y_t \sim I(1)$ if $\Delta Y_t = Y_t - Y_{t-1} \sim I(0)$
  - Most common in economics: prices, exchange rates, GDP levels
  
- **I(d)**: General case requiring $d$ differences

**Economic Interpretation:**
- **I(0)**: Returns, growth rates, inflation (fluctuate around a constant mean)
- **I(1)**: Prices, exchange rates, GDP levels (wander without bound)

---

### **11.2 Random Walk Models**

The simplest **I(1)** process is the **random walk**:

**Random Walk (No Drift):**
$$Y_t = Y_{t-1} + \varepsilon_t, \quad \varepsilon_t \sim \text{i.i.d.}(0, \sigma^2)$$

**Properties:**
- $Y_t = Y_0 + \sum_{i=1}^{t} \varepsilon_i$ (cumulative sum of shocks)
- $\mathbb{E}[Y_t] = Y_0$ (constant mean)
- $\text{Var}(Y_t) = t\sigma^2$ ✗ **Non-stationary** (variance grows with time!)
- $\Delta Y_t = \varepsilon_t \sim I(0)$ ✓ **First difference is stationary**

**Random Walk with Drift:**
$$Y_t = \mu + Y_{t-1} + \varepsilon_t$$

**Properties:**
- $Y_t = Y_0 + \mu t + \sum_{i=1}^{t} \varepsilon_i$
- Mean: $\mathbb{E}[Y_t] = Y_0 + \mu t$ ✗ **Non-stationary** (trending mean)
- Variance: $\text{Var}(Y_t) = t\sigma^2$ ✗ **Non-stationary**
- $\Delta Y_t = \mu + \varepsilon_t \sim I(0)$ ✓ **First difference is stationary around drift**

---

### **11.3 Deterministic vs Stochastic Trends**

**Deterministic Trend:**
$$Y_t = \beta_0 + \beta_1 t + \varepsilon_t, \quad \varepsilon_t \sim I(0)$$

- Trend is **fixed** (deterministic function of time)
- Removing trend: $Y_t - \beta_1 t = \beta_0 + \varepsilon_t$ is stationary
- **Trend-stationary**: I(0) after detrending

**Stochastic Trend (Unit Root):**
$$Y_t = Y_{t-1} + \varepsilon_t = Y_0 + \sum_{i=1}^{t} \varepsilon_i$$

- Trend is **random** (cumulative sum of shocks)
- Removing deterministic trend **does not** make it stationary
- **Difference-stationary**: I(1), requires differencing

**Key Distinction:**
- **Deterministic trend**: Detrend → stationary
- **Stochastic trend**: Difference → stationary

**Why It Matters:**
- **Forecasting**: Deterministic trends predictable; stochastic trends not
- **Spurious regression**: Two I(1) series can appear correlated even if independent
- **Cointegration**: Two I(1) series may share a common stochastic trend

---

### **11.4 Visual Intuition: EUR/USD Exchange Rate**

Let's examine the EUR/USD exchange rate to see non-stationarity in action.